tools/lsp/language.rs

```diff
@@ -576,33 +596,185 @@ pub fn show_preview_command(
     Ok(())
 }
 
+fn populate_command_range(node: &SyntaxNode) -> Option<lsp_types::Range> {
+    let range = node.text_range();
+
+    let start_offset = node
+        .text()
+        .find_char('\u{0002}')
+        .and_then(|s| s.checked_add(1.into()))
+        .unwrap_or(range.start());
+    let end_offset = node.text().find_char('\u{0003}').unwrap_or(range.end());
+
+    (start_offset <= end_offset).then_some(util::text_range_to_lsp_range(
+        &node.source_file,
+        TextRange::new(start_offset, end_offset),
+    ))
+}
+
+pub async fn populate_command(
+    params: &[serde_json::Value],
+    ctx: &Rc<Context>,
+) -> Result<serde_json::Value, LspError> {
+    let text_document =
+        serde_json::from_value::<lsp_types::OptionalVersionedTextDocumentIdentifier>(
+            params
+                .first()
+                .ok_or_else(|| LspError {
+                    code: LspErrorCode::InvalidParameter,
+                    message: "No textdocument provided".into(),
+                })?
+                .clone(),
+        )
+        .map_err(|_| LspError {
+            code: LspErrorCode::InvalidParameter,
+            message: "First paramater is not a OptionalVersionedTextDocumentIdentifier".into(),
+        })?;
+    let new_text = serde_json::from_value::<String>(
+        params
+            .get(1)
+            .ok_or_else(|| LspError {
+                code: LspErrorCode::InvalidParameter,
+                message: "No code to insert".into(),
+            })?
+            .clone(),
+    )
+    .map_err(|_| LspError {
+        code: LspErrorCode::InvalidParameter,
+        message: "Invalid second parameter".into(),
+    })?;
+
+    let edit = {
+        let document_cache = &mut ctx.document_cache.borrow_mut();
+        let uri = text_document.uri;
+        let version = document_cache.document_version(&uri);
+
+        if let Some(source_version) = text_document.version {
+            if let Some(current_version) = version {
+                if current_version != source_version {
+                    return Err(LspError {
+                        code: LspErrorCode::InvalidParameter,
+                        message: "Document version mismatch. Please refresh your command data"
```

Comment:
Who is the "reciever" of the message? It's the user, right? I wouldn't have any idea of "Please refresh your command data" means

Reply:
I removed that part.